### DISCUSSION TITLE
Stiffening Members get converted into Members

### DISCUSSION THREAD

**Discussion by tbyrk** [idea-statica/ideastatica-public#491] [Rhino/Grasshopper] [answered]
The script is based on the example "CON_Push new operations into an existing Connection file". When I push one plate into my base connection the Stiffening Members turn into actual Members with wrong pitches. Seems like Stiffening Members are recognized by Grasshopper components as regular Members.

Second issue is operations. Could I add a plate into that connection without touching any other operations? It could just be one stiffening plate operation. Right now all operations get overwritten by something else.

<img width="1278" alt="Stiffening Member" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/6db46430-d487-49bc-9045-f18846dedf4d">

<img width="1279" alt="Not a Member 1" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/c3f8ae2b-8cf3-40ae-bea9-eb8459521d26">

<img width="1278" alt="Not a Member 2" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/3f46dd07-56fc-46fd-9c56-16e0f3cb2703">

**Comment by nathanlukeIDEA** [comment]
Hi @tbyrk, Can you please send me your files?  

I would anticipate that the Members would be shown there as stiffening members are still defined as Members in IOM, however, I could hide them from this list, which I understand does make sense from your point of view.  

However, they should not get overridden as members when updating the connection. Let me take a look at trying to fix this.

_Surgically injecting a plate_, is not something we can do atm. I know this sounds like a good approach, but once you inject it you will need to access all the operation properties to define relations to other elements such as cuts welds, bolts etc. 

In the first instance, I would look at seeing if we can significantly improve the update functionality to not overwrite your existing operations, members or loadings.

**Comment by tbyrk** [comment]
Zip:
- Add plates to existing IDEA Connection.gh
- Pipes_v2.ideaCon
- Pipes_v2_updated.ideaCon
- Pipes_stub.ideaCon
- Pipes_stub_updated.ideaCon
- Plate-for-Nathan.3dm

The purpose of this script is to get auxiliary lines according to member local axes so you can bake them and then easily draw custom plate geometry in Rhino to push into the connection file.

I added a connection with "stub" operations which also gets messed up.
<img width="1719" alt="Connection with stubs" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/16cf6ca7-1841-4c9f-99a6-9ac10316c8ab">

Thanks for taking a look!

[ForNathan.zip](https://github.com/idea-statica/ideastatica-public/files/14753842/ForNathan.zip)

**Comment by nathanlukeIDEA** [comment, marked as answer]
Hi @tbyrk, 

Thanks for your patience on this one.

1) I have fixed the update (0.1.5) to now ensure that the stubs are recreated as stub elements when updating.

This workflow only really works well when there are no existing operations in the connection.  This is something that we want to improve so thanks for the great example.

HOWEVER, there is a workaround to achieve what you are after. It is not that pretty but works pretty well. I have attached a .zip with a script which will do it for you. You will need to export the original connection as a template - I can automate this through the Grasshopper plugin if it helps, but I need to add the component.  

The idea is like this:
1) Export a template of the original connection (need to do this in the dev tab).
![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/646faca2-0812-4d67-9bc6-e7e52880c472)

2) In Grasshopper, reference the existing connection. Create a new connection point that references the members (basically creating a blank connection point). If it has the same **name** it will replace the connection point with all the existing operations in it.
![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/703dc4ef-11e6-4eff-bc8a-2ef73397777f)

3) Update the referenced connection with only the plate in it. We remove all existing operations before doing this. You will get something like this:
![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/7faad50c-5b9b-42ce-bbd8-24cf866c4046)

4) Apply the original template back to the connection with the plate in it. When applying the template we do not reset the operations to keep the plate in there. 
![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/0c7718fb-300b-419c-b676-0c24ca4ce5e5)

And you will get the original connection operations imported with the plate in it. :)
![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/92686a35-1672-4c36-afb9-e7ea104f4a2d)

Hope this helps. 

Cheers 
Nathan

[inject_plate.zip](https://github.com/idea-statica/ideastatica-public/files/14900007/inject_plate.zip)

**Comment by tbyrk** [comment]
Thanks, Nathan! Worked well.

Now that I know it is possible for the stiffening member to remain in a connection through the "apply template" workflow I want to test one more thing. I'd like to add a stiffening member with a vector value that is the average of all member vector values. Not sure if this even works. Is it even possible to use the average vector to modify the stiffening member direction vector parameters? Files attached.

<img width="1287" alt="image" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/aa497891-b03f-495e-b458-05c1a7001f94">

<img width="1279" alt="image" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/04886467-27db-4152-b935-971bb48e2706">

<img width="894" alt="Average 1" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/5ca88a6d-2254-4b57-8538-1a42c79690c3">
[Stiffening_member_alignment.zip](https://github.com/idea-statica/ideastatica-public/files/14903640/Stiffening_member_alignment.zip)

<img width="1274" alt="Average 2" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/7bd5e887-b02d-466c-b2b5-7ee733405292">

**Comment by tbyrk** [comment]
Vector issue solved. I think I can manage the rest.

What is that error though? It works with the error.

<img width="202" alt="image" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/8f64de1a-d4ec-49fd-a1cc-ca90d1d305a4">

<img width="1277" alt="Script" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/8d088735-7a84-432e-a3c3-1e73ad2ef564">

<img width="1902" alt="Modified connection" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/cbe00961-8596-460b-bc3d-d23e84f80109">

[Rotate_Stiffening_Member.zip](https://github.com/idea-statica/ideastatica-public/files/14915928/Rotate_Stiffening_Member.zip)

**Comment by nathanlukeIDEA** [comment]
Hi @tbyrk, the parameter is expecting the value to be wrapped in quotes. In grasshopper if you add single quoate wrappers it should work without error.

**Comment by tbyrk** [comment]
Works, thanks!

<img width="818" alt="image" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/d5f65c18-be1d-41bd-a192-25546a51d9f7">

**Comment by tbyrk** [comment]
I only now realized that "Stiffening Member" component has been added. No need for templates.

<img width="1142" alt="SM1" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/10f5a6cb-aa32-4417-9445-19c102322d32">

<img width="926" alt="SM2" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/539c7a22-50cc-4071-8f1a-85b5f579410e">

**Reply by nathanlukeIDEA** [reply]
Yes, but rotating the existing stiffening member vector is a valid use case, so it is great you got it working. Looking forward to seeing the finished product. Please put it in the show and tell :) https://github.com/idea-statica/ideastatica-public/discussions/categories/show-and-tell

**Comment by tbyrk** [comment]
Bulk connection generation actually started to work. Can't believe it. It is a funny connection but proof of concept is what matters for me right now. All connections have their own vector pointing towards the center of the dome.

<img width="1022" alt="image" src="https://github.com/idea-statica/ideastatica-public/assets/162557189/5ae7194b-9560-4a28-a0bd-130ce7daadc6">
